docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+# Writing your own Custom Middleware
+
+## How to write your own custom middleware for Mirascope
+
+`middleware_decorator` is a helper function to assist in helping you wrap any Mirascope call.
+We will be creating an example decorator `with_saving` that saves some metadata after a Mirascope call:
+
+### Writing the decorator
+
+```python
+from mirascope.integrations import middleware_decorator
+
+def with_saving(fn):
+    """Saves some data after a Mirascope call."""
+
+    return middleware_decorator(
+        fn,
+        custom_context_manager=custom_context_manager,
+        custom_decorator=custom_decorator,
+        handle_call_response=handle_call_response,
+        handle_call_response_async=handle_call_response_async,
+        handle_stream=handle_stream,
+        handle_stream_async=handle_stream_async,
+        handle_response_model=handle_response_model,
+        handle_response_model_async=handle_response_model_async,
+        handle_structured_stream=handle_structured_stream,
+        handle_structured_stream_async=handle_structured_stream_async,
+    )
+```
+
+We will go over each of the different functions.
+
+### `handle_call_response` and `handle_call_response_async`
+
+These functions will be called after making a mirascope call with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base import BaseCallResponse
+
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call here
+```
+
+The first argument will be a Mirascope `CallResponse` of the provider you are using.
+The second argument will be your Mirascope call function.
+The third argument is the yielded object of your `custom_context_manager`
+
+`handle_call_response_async` is the same as `handle_call_response` but using an `async` function.
+
+### `handle_stream` and `handle_stream_async`
+
+These functions will be called after streaming a Mirascope call with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base._stream import BaseStream
+
+
+def handle_stream(
+    stream: BaseStream, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after stream here
+```
+
+The first argument will be a Mirascope `Stream` of the provider you are using.
+See `handle_call_response` section for information regarding the other arguments.
+
+One thing to note for `handle_stream` is that it will not be called until after the `Generator` has been exhausted.
+
+### `handle_response_model` and `handle_response_model_async`
+
+These functions will be called after making a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+from pydantic import BaseModel
+
+from mirascope.core.base._utils._base_type import BaseType
```

Comment:
hmm, I don't like telling people to import something from a private module. Instead this should be:

```python
from mirascope.core.base import BaseType
```

This will require updating the internal imports in the init, and that's fine.